src/api/images/images.service.ts

```diff
@@ -0,0 +1,72 @@
+import { Injectable, InternalServerErrorException, BadRequestException } from '@nestjs/common';
+import * as AWS from 'aws-sdk';
+import { v4 as uuidV4 } from 'uuid';
+import  * as multer from 'multer';
+
+const ONE_MEGABYTE: number = 1024 * 1024;
+const JPEG_MIME_TYPE = 'image/jpeg';
+const PNG_MIME_TYPE = 'image/png';
+
+@Injectable()
+export class ImagesService {
+
+    private readonly s3 = new AWS.S3({
+        endpoint: process.env.DIGITALOCEAN_SPACES_ENDPOINT,
+        accessKeyId: process.env.DIGITALOCEAN_SPACES_ACCESS_KEY,
+        secretAccessKey: process.env.DIGITALOCEAN_SPACES_SECRET,
+    });
+
+    /**
+     * Uploads the given form-data for the user with the given ID. On success,
+     * returns the full URL of the uploaded image.
+     * @throws(BadRequestException) if the image is larger than 3MB, or is not a PNG or JPEG.
+     * @throws(InternalServerException) if the upload to Spaces fails.
+     * @returns The full URL of the uploaded image. 
+     */
+    async upload(imageFile: Express.Multer.File, userId: string): Promise<string> {
+        this.validateImageFile(imageFile);
+
+        const extension = this.getExtension(imageFile);
+
+        const uploadOptions: AWS.S3.PutObjectRequest = {
+            Bucket: process.env.DIGITALOCEAN_SPACES_NAME,
+            Key: `${userId}-${uuidV4()}.${extension}`, // Note: Keys must be unique per-Space
+            Body: imageFile.buffer,
+            ACL: 'public-read',
+            ContentType: imageFile.mimetype,
+        };
+
+        return new Promise((resolve) => {
+            this.s3.upload(uploadOptions, null, (err: Error, data: AWS.S3.ManagedUpload.SendData) => {
+                if (err) {
+                    throw new InternalServerErrorException(err);
```

Comment:
This _will_ throw out to callers if they've awaited this method, right?

Reply:
Yup. Any time a NestJS exception is thrown, no matter where it is, the function it's in will error out.